src/config.rs

```diff
@@ -0,0 +1,376 @@
+// Copyright 2024 Globus Medical, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+use miette::{bail, ensure, miette, Context, IntoDiagnostic};
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+    str::FromStr,
+};
+
+use crate::{
+    manifest::{Edition, CANARY_EDITION},
+    registry::{RegistryRef, RegistryUri},
+};
+
+/// Location of the configuration file
+const CONFIG_FILE: &str = ".buffrs/config.toml";
+
+/// Key for common default arguments
+const DEFAULT_ARGS_KEY: &str = "*";
+
+/// Representation of the configuration file
+///
+/// # Example
+///
+/// ```toml
+/// edition = "0.10.0"
+///
+/// [registries]
+/// some_org = "https://artifactory.example.com/artifactory/some-org"
+///
+/// [registry]
+/// default = "some_org"
+///
+/// [commands]
+/// default_args = ["--insecure"]
+///
+/// [commands.install]
+/// default_args = ["--generate-buf-yaml"]
+/// ```
+///
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Config {
+    /// Edition of this configuration file (in sync with the Proto.toml edition)
+    edition: Edition,
+
+    /// Path to the configuration file
+    config_path: Option<PathBuf>,
+
+    /// Default registry alias to use if none is specified
+    default_registry: Option<String>,
+
+    /// List of registries
+    registries: HashMap<String, RegistryUri>,
+
+    /// Default arguments for commands
+    command_defaults: HashMap<String, Vec<String>>,
+}
+
+impl Config {
+    /// Create a new configuration with default values
```

Comment:
This comment lies 

Could you roughly follow the public api semantics from the manifest / cache?

Reply:
Can you give me specifics?

Reply:
Ah yes sorry.

So this function will do IO although the comment says it will create a new one with default values!

With regards to the common pattern I was referring to the `Manifest` struct for example

Reply:
24136669c705663dd0845c21ca7e77259d55eff8